application/positionmanager/src/positionmanager/main.py

```diff
@@ -0,0 +1,285 @@
+
+        data = pl.DataFrame(response_data["data"])
+
+        data = data.with_columns(
+            pl.col("timestamp")
+            .str.slice(0, 10)
+            .str.strptime(pl.Date, "%Y-%m-%d")
+            .alias("date")
+        )
+
+        if "ticker" not in data.columns or "close_price" not in data.columns:
+            raise ValueError("Data must contain ticker and close_price columns")
+
+        data = (
+            data.sort("date")
+            .pivot(index="date", columns="ticker", values="close_price")
+            .with_columns(pl.all().cast(pl.Float64))
+        )
+
+        return data
+
+
+class PortfolioOptimizer:
+    def __init__(
+        self,
+        minimum_portfolio_tickers: int = 5,
+        maximum_portfolio_tickers: int = 20,
+    ):
+        self.minimum_portfolio_tickers = minimum_portfolio_tickers
+        self.maximum_portfolio_tickers = maximum_portfolio_tickers
+
+    def get_optimized_portfolio(
+        self,
+        data: pl.DataFrame,
+        portfolio_value: int,
+    ) -> Dict[str, float]:
+        converted_data = data.to_pandas()
+
+        if "date" in converted_data.columns:
+            converted_data = converted_data.set_index("date")
+
+        mu = expected_returns.mean_historical_return(converted_data, frequency=252)
+
+        S = risk_models.CovarianceShrinkage(converted_data).ledoit_wolf()
+
+        long_only_weight_bounds = (0, 1)
+        efficient_frontier = EfficientFrontier(
+            mu, S, weight_bounds=long_only_weight_bounds
+        )
+
+        efficient_frontier.max_sharpe(risk_free_rate=0.02)  # 2% risk-free rate
+        weights = efficient_frontier.clean_weights()
+
+        latest_prices = get_latest_prices(converted_data)
+
+        discrete_allocation = DiscreteAllocation(
+            weights, latest_prices, total_portfolio_value=portfolio_value
+        )
+
+        optimized_portfolio, _ = discrete_allocation.greedy_portfolio()
+
+        return optimized_portfolio
+
+
+trading_days_per_year = 252
+
+application = FastAPI()
+
+
+def get_alpaca_client():
+    return AlpacaClient(
+        api_key=os.getenv("ALPACA_API_KEY"),
+        api_secret=os.getenv("ALPACA_API_SECRET"),
+        paper=os.getenv("ALPACA_PAPER", "true").lower() == "true",
+    )
+
+
+def get_data_client():
+    return DataClient(datamanager_base_url=os.getenv("DATAMANAGER_BASE_URL"))
+
+
+def get_portfolio_optimizer():
+    return PortfolioOptimizer(
+        minimum_portfolio_tickers=int(os.getenv("MINIMUM_PORTFOLIO_TICKERS", "5")),
+        maximum_portfolio_tickers=int(os.getenv("MAXIMUM_PORTFOLIO_TICKERS", "20")),
+    )
+
+
+@application.get("/health")
+def get_health():
+    return {"status": "healthy"}
+
+
+@application.post("/positions")
+def create_position(payload: PredictionPayload) -> Dict[str, Any]:
+    alpaca_client = get_alpaca_client()
+    data_client = get_data_client()
+    portfolio_optimizer = get_portfolio_optimizer()
```

Comment:
if these are single getter without inputs, I would just put them directly into this function, if they are just wrapping up environment variables, I wouldn't worry that much about unit testing them, and I would say to put them in like an app.state to share across

Reply:
I opted just to stick them in the handlers directly even though that's sub-optimal.